File: scala-project/src/main/scala/com/waitingforcode/Reader.scala
package com.waitingforcode

import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, functions => F}

object Reader {

  def selectRawVisits(visitsSource: DataFrame): DataFrame = {
    val visitSchema: StructType =
      StructType.fromDDL("""
        |visit_id STRING, event_time TIMESTAMP, user_id STRING, page STRING,
        |        context STRUCT<
        |            referral STRING, ad_id STRING,
        |            user STRUCT<
        |                ip STRING, login STRING, connected_since TIMESTAMP
        |            >,
        |            technical STRUCT<
        |                browser STRING, browser_version STRING, network_type STRING, device_type STRING, device_version STRING
        |            >
        |        >
        |""".stripMargin)

    visitsSource.select(F.from_json(
        F.col("value").cast("string"), visitSchema,
      Map("timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ssXXX")).as("value")
    ).selectExpr("value.*")
  }

}

File: python-project/visits_generator/reader.py
from pyspark.sql import functions as F, DataFrame


def select_raw_visits(visits_source: DataFrame) -> DataFrame:
    # the schema doesn't contain all fields; instead it focuses on the required attributes
    # for the session generation
    visit_schema = '''
        visit_id STRING, event_time TIMESTAMP, user_id STRING, page STRING,
        context STRUCT<
            technical STRUCT<
                device_type STRING, device_version STRING
            >
        >
    '''
    visits_raw = (visits_source.select(
        F.from_json(F.col('value').cast('string'), visit_schema).alias('value'))
            .selectExpr('value.*'))

    # Exploding the context columns to make it easier to query later
    return visits_raw.withColumns({'device_type': visits_raw.context.technical.device_type,
                                   'device_version': visits_raw.context.technical.device_version}).drop('context')
